crates/environ/src/module_types.rs

```diff
@@ -323,27 +325,94 @@ where
+                // array vs struct vs func reference. In this case, we can use
+                // the validator's type context.
+                if let Some(ty) = self.types.types.wasm_types.get(interned) {
+                    match &ty.composite_type {
+                        WasmCompositeType::Array(_) => WasmHeapType::ConcreteArray(index),
+                        WasmCompositeType::Func(_) => WasmHeapType::ConcreteFunc(index),
+                    }
+                } else if let Some((wasmparser_types, _)) = self.rec_group_context.as_ref() {
+                    match &wasmparser_types[id].composite_type {
+                        wasmparser::CompositeType::Array(_) => WasmHeapType::ConcreteArray(index),
+                        wasmparser::CompositeType::Func(_) => WasmHeapType::ConcreteFunc(index),
+                        wasmparser::CompositeType::Struct(_) => unreachable!(),
+                    }
+                } else {
+                    panic!("forward reference to type outside of rec group?")
+                }
             }
+
             UnpackedIndex::Module(module_index) => {
                 let module_index = TypeIndex::from_u32(module_index);
-                let interned_index = self.module.types[module_index];
-                WasmHeapType::ConcreteFunc(EngineOrModuleTypeIndex::Module(interned_index))
+                let interned = self.module.types[module_index];
+                let index = EngineOrModuleTypeIndex::Module(interned);
+
+                // See comment above about `wasm_types` maybe not having the
+                // converted sub type yet. However in this case we don't have a
+                // `wasmparser::types::CoreTypeId` on hand, so we have to
+                // indirectly get one by looking it up inside the current rec
+                // group.
+                if let Some(ty) = self.types.types.wasm_types.get(interned) {
+                    match &ty.composite_type {
+                        WasmCompositeType::Array(_) => WasmHeapType::ConcreteArray(index),
+                        WasmCompositeType::Func(_) => WasmHeapType::ConcreteFunc(index),
+                    }
+                } else if let Some((parser_types, rec_group)) = self.rec_group_context.as_ref() {
+                    let rec_group_index = interned.index() - self.types.types.wasm_types.len();
+                    let id = parser_types
+                        .rec_group_elements(*rec_group)
+                        .nth(rec_group_index)
+                        .unwrap();
+                    match &parser_types[id].composite_type {
+                        wasmparser::CompositeType::Array(_) => WasmHeapType::ConcreteArray(index),
+                        wasmparser::CompositeType::Func(_) => WasmHeapType::ConcreteFunc(index),
+                        wasmparser::CompositeType::Struct(_) => unreachable!(),
+                    }
+                } else {
```

Comment:
Yeah I attempted this and then aborted when I saw how it was shaping up. I think factoring things out so that you can still capture the differences with a function or trait or enum or set of option arguments or something would ultimately end up with the same amount of code but slightly more convoluted because now you have to understand how the different cases work out with the abstraction mechanism.